src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,256 @@
+  rwa [mul_assoc, mul_assoc, eq_inv_mul_iff_mul_eq, ←mul_assoc, ←mul_assoc, eq_mul_inv_iff_mul_eq],
+end
+
+lemma disjoint_doset (H K : subgroup G) (a b : G) (h: ¬ disjoint (doset a H K) (doset b H K)) :
+  doset a H K = doset b H K :=
+begin
+  rw disjoint.comm at h,
+  have ha : a ∈ (doset b H K), by {apply mem_doset_of_not_disjoint _ _ _ _ h},
+  apply doset_eq_of_mem ha,
+end
+
+/--Create a doset out of an element of `H \ G / K`-/
+def quot_to_doset (H K : subgroup G) (q : quotient ↑H ↑K ) : set G := (doset q.out' H K)
+
+/--Map from `G` to `H \ G / K`-/
+abbreviation mk (H K : subgroup G) (a : G) : quotient ↑H ↑K :=
+quotient.mk' a
+
+instance (H K : subgroup G) : inhabited (quotient ↑H ↑K) := ⟨(mk H K (1 : G) : quotient ↑H ↑K)⟩
+
+lemma eq (H K : subgroup G) (a b : G): mk H K a = mk H K b ↔ ∃ (h ∈ H) (k ∈ K), b = h * a * k :=
+by { rw quotient.eq', apply (rel_iff ), }
+
+lemma out_eq' (H K : subgroup G) (q : quotient ↑H ↑K) : mk H K q.out' = q :=
+quotient.out_eq' q
+
+lemma mk_out'_eq_mul (H K : subgroup G) (g : G) :
+  ∃ (h k : G), (h ∈ H) ∧ (k ∈ K) ∧ (mk H K g : quotient ↑H ↑K).out' = h * g * k :=
+begin
+  have := eq H K (mk H K g : quotient ↑H ↑K).out' g,
+  rw out_eq' at this,
+  simp only [exists_prop] at this,
+  have h: mk H K g = mk H K g, by {refl,},
+  have hh := this.1 h,
+  obtain ⟨h,h_h,k,hk,T⟩ :=hh,
+  refine ⟨h⁻¹ ,k⁻¹, (H.inv_mem h_h), K.inv_mem hk,_⟩,
+  rw T,
+  simp_rw ← mul_assoc,
+  simp only [one_mul, mul_left_inv, mul_inv_cancel_right],
+  exact congr_arg quotient.out' (congr_arg (mk H K) (eq.symm T)),
+end
+
+lemma doset_eq_quot_eq (H K : subgroup G) (a b : G) (h : doset a H K = doset b H K) :
+  mk H K a = mk H K b :=
+begin
+  rw eq,
+  have : b ∈ doset a H K,
+  { rw h,
+    simp only [exists_prop, set_coe.exists, mem_doset, set_like.mem_coe, subgroup.coe_mk,
+    prod.exists],
+    use 1,
+    simp only [H.one_mem, K.one_mem, one_mul, exists_eq_right, self_eq_mul_right, and_self]},
+  rw mem_doset at this,
+  simp only [exists_prop, set_coe.exists, set_like.mem_coe, subgroup.coe_mk, prod.exists] at *,
+  apply this,
+end
+
+lemma disjoint_doset' (H K : subgroup G) (a b : quotient H.1 K) :
+  a ≠ b → disjoint (doset a.out' H K) (doset b.out' H K) :=
+begin
+  simp only [ne.def],
+  contrapose,
+  simp only [not_not],
+  intro h,
+  have := disjoint_doset H K _ _ h,
+  have h2 := doset_eq_quot_eq _ _ _ _ this,
+  simp_rw [out_eq'] at h2,
+  apply h2,
+end
+
+lemma top_eq_union_dosets (H K : subgroup G) : (⊤ : set G) = ⋃ q, quot_to_doset H K q :=
+begin
+  simp only [set.top_eq_univ],
+  ext,
+  simp only [set.mem_Union, true_iff, set.mem_univ],
+  use mk H K x,
+  rw quot_to_doset,
+  simp only [exists_prop, mem_doset, subgroup.mem_carrier, set_like.mem_coe],
+  have hy := mk_out'_eq_mul H K x,
+  rcases hy with ⟨h, k, h3, h4, h5⟩,
+  refine ⟨h⁻¹, H.inv_mem h3 ,k⁻¹, K.inv_mem h4,_⟩,
+  simp only [h5, subgroup.coe_mk, ←mul_assoc, one_mul, mul_left_inv, mul_inv_cancel_right],
```

Comment:
`ext, simp [set.mem_univ, true_iff]` can be simplified to `set.eq_univ_of_forall`, and `have, rcases this` has an abbreviation `obtain`:
```suggestion
  simp only [set.mem_Union, quot_to_doset, mem_doset],
  use mk H K x,
  obtain ⟨h, k, h3, h4, h5⟩ := mk_out'_eq_mul H K x,
  refine ⟨h⁻¹, H.inv_mem h3, k⁻¹, K.inv_mem h4, _⟩,
  simp only [h5, subgroup.coe_mk, ←mul_assoc, one_mul, mul_left_inv, mul_inv_cancel_right],
```

Reply:
I couldn't quite get the same proof you had to work.